src/rust/lib/graph-editor/src/component/edge.rs

```diff
@@ -44,23 +47,280 @@ const NODE_PADDING       : f32 = node::SHADOW_SIZE;
+        let local          = self.to_shape_coordinate_system(point);
+        let local_snapped  = self.snap_to_shape_local(local)?;
+        let global_snapped = self.from_shape_to_global_coordinate_system(local_snapped);
+        Some(global_snapped)
+    }
+
+    /// Set the hover split for this shape. The `split_data` indicates where the shape should be
+    /// split and which side should be recolored.
+    fn enable_hover_split(&self, split_data:SplitData) {
+        let SplitData {split_position,area,cut_angle} = split_data;
+        // Compute rotation in shape local coordinate system.
+        let base_rotation        = self.display_object().rotation().z;
+        let hover_split_rotation = cut_angle + base_rotation;
+        match area {
+            Area::Above => self.set_hover_split_rotation(hover_split_rotation),
+            Area::Below => {
+                self.set_hover_split_rotation(hover_split_rotation + 2.0 * RIGHT_ANGLE)
+            },
+        }
+        // Compute position in shape local coordinate system.
+        let offset = self.to_shape_coordinate_system(split_position);
+        self.set_hover_split_offset(offset)
+    }
+
+    /// Disable the hover split on this shape.
+    fn disable_hover_split(&self) {
+        self.set_hover_split_offset(Vector2::new(INFINITE, INFINITE));
+        self.set_hover_split_rotation(RIGHT_ANGLE);
+    }
+
+    /// Convert the given global coordinates to the shape local coordinate system.
+    fn to_shape_coordinate_system(&self, point:Vector2<f32>) -> Vector2<f32> {
+        let base_rotation   = self.display_object().rotation().z;
+        let local_unrotated = point - self.display_object().global_position().xy();
+        UnitComplex::new(-base_rotation) * local_unrotated
+    }
+
+    /// Convert the given shape local coordinate to the global coordinates system.
+    fn from_shape_to_global_coordinate_system(&self, point:Vector2<f32>) -> Vector2<f32> {
+        let base_rotation   = self.display_object().rotation().z;
+        let local_unrotated = UnitComplex::new(base_rotation) * point;
+        local_unrotated + self.display_object().global_position().xy()
+    }
+
+    /// Check whether the given shape local coordinate is within the bounds of the shape.
+    fn contains_local(&self, point:Vector2<f32>) -> bool {
+        let size = self.sprite().size.get();
+        if point.x < -size.x / 2.0 || point.y < -size.y / 2.0 {
+            return false
+        }
+        point.x < size.x / 2.0 && point.y < size.y / 2.0
+    }
+
+    /// Check whether the given global coordinate is within the bounds of the shape.
+    fn contains(&self, point:Vector2<f32>) -> bool {
+        self.contains_local(self.to_shape_coordinate_system(point))
+    }
+}
+
+
+
+// ========================
+// === Edge Shape Trait ===
+// ========================
+
+/// Indicates which area should be recolored.
+#[derive(Clone,Copy,Debug,Eq,PartialEq)]
+enum Area {
+    Above,
+    Below,
+}
```

Comment:
This needs better name. Moreover, I'm not sure if above / below are good names, especially in context of upwards edges. I would name it more like `split::Side::Source` and `split::Side::Target`, what do you think about it? Of course, I'm open to better names! If you don't want to create separate module for split.yet, then `SplitSide::Source` etc